### PR TITLE
GODRIVER-2161 Add NewSingleResultFromDocument and NewCursorFromDocuments functions

### PR DESCRIPTION
**Pull request by benjirewis**
GODRIVER-2161

Adds a new `SingleResult` constructor `NewSingleResultFromDocument` that makes a `SingleResult` with an underlying `Cursor` pre-loaded with a singleton slice of documents (`[]interface{}{document}`) as the first batch. Adds a new `Cursor` construct `NewCursorFromDocuments` that makes a `Cursor` pre-loaded with a slice of documents (`[]interface{}`) as the first batch. These new functions can be used to mock server responses to read operations (e.g. `Find`, `FindOne`, `Aggregate`). Adds associated unit and integration tests.

### CODE REVIEW

File: mongo/single_result.go [hunk 1]
```diff
@@ -28,6 +29,22 @@ type SingleResult struct {
 	reg *bsoncodec.Registry
 }
 
+// NewSingleResultFromBytes creates a SingleResult with an underlying Cursor pre-loaded with the provided contents.
+func NewSingleResultFromBytes(contents []byte) *SingleResult {
```
Comment (benjirewis) [inline mongo/single_result.go:33]: I opted to only add `NewSingleResultFromBytes` and _not_ `NewCursorFromBytes`. The latter seemed unnecessary to mock responses to read operations. Unlike `BatchCursor`, `Cursor` and `SingleResult` are both under the `mongo` package, so you can just as easily access `Cursor` internals from `single_result.go`. Let me know if you think users would benefit from `NewCursorFromBytes`.

File: mongo/single_result.go [hunk 2]
```diff
@@ -28,6 +29,22 @@ type SingleResult struct {
...
+	c := &Cursor{
+		bc:       driver.NewBatchCursorFromBytes(contents),
+		registry: bson.DefaultRegistry,
```
Comment (benjirewis) [inline mongo/single_result.go:36]: We could let users specify a custom `registry` here.

File: mongo/single_result.go [hunk 3]
```diff
@@ -71,6 +88,12 @@ func (sr *SingleResult) setRdrContents() error {
 		return nil
 	case sr.cur != nil:
 		defer sr.cur.Close(context.TODO())
+
+		// Set contents of rdr to cur.Current if non-nil. Otherwise, iterate underlying
+		// cursor.
+		if sr.rdr = sr.cur.Current; sr.rdr != nil {
```
Comment (benjirewis) [inline mongo/single_result.go:94]: Most cursors start with an empty first batch and must be iterated with `Next` to update `Current`. `Cursor`s created with `NewSingleResultFromBytes` start with a pre-loaded `Current`, so we don't need to iterate. This new block shouldn't affect regular `SingleResult`'s rdr-setting logic, but let me know if you think otherwise.

**Comment by SVilgelm** [comment]
really, so many problems with testing projects that uses mongo. You definitely need to spend a couple of weeks and implements interfaces for all your types and bump v2. You will be really surprised how fast the customers will switch to new version to get an ability of mocking the mongo interfaces.
but now you are spending the time on hacking the cursors and etc...

**Comment by benjirewis** [comment]
Hello @SVilgelm ! We appreciate the feedback, and we've indeed received a number of requests to make the main API types (`Client`, `Database` and `Collection`) interfaces instead of structs. 

Unfortunately, we will not be doing so anytime soon. The main hesitation around exposing an interface is backwards compatibility. For struct types, the only breaking changes are _removing_ existing functions or _changing_ function signatures. We're able to add new functions whenever we want (which is somewhat frequently, given our need to keep up with MongoDB server features and other user requests). For interfaces, though, even _adding_ functions is a breaking change because we are technically breaking all external implementations by doing so. Even if we switched to interfaces in a 2.0.0 of the driver, we'd be unable to alter those interfaces from that point on. That's not a feasible path forward for us at the moment.

**Comment by SVilgelm** [comment]
@benjirewis thank you for the response. I totally understand all possible problems. But I would really really ask you, if you are going to release v2 of the driver with backward incompatible changes, please create the interfaces for all types.

> For interfaces, though, even adding functions is a breaking change because we are technically breaking all external implementations by doing so

that's true, but the main reason of having the interfaces is creating mocks and using them for testing, so if you would change an interface by adding new function, then we (customers) will just regenerate the mocks. Of course I understand that someone can use the interfaces differently, but they will need to spend a bit more time to adjust their code. 

> For struct types, the only breaking changes are removing existing functions or changing function signatures. 

in this case you must to bump new major release anyway, so you will need to change the interfaces as well

> (Client, Database and Collection)

actually not only those 3, but all other as well, you can check my implementation for full list on needed interfaces: https://github.com/sv-tools/mongoifc#interfaces


The main concern I have about your position is that your response for such request is always different, probably depends on who's is asking. There are closed issues in Jira with won't do and different responses, but now when the Coinbase requested a hack you are implementing it. Why did not you answer for this request like you did for another? When you recommended having an internal DB layer or setting up a real mongo for integration tests?

File: mongo/single_result.go (hunk 1 above)
Comment (matthewdale) [reply mongo/single_result.go:33]: I'm not sure I follow the rationale for not needing a way to create a `Cursor`. Consider the example use case for testing a set of functions that need to call `FindOne` and `Find` by defining a mock type that satisfies that interface:
```go
type finder interface {
	Find(context.Context, interface{}, ...*options.FindOptions) (*mongo.Cursor, error)
	FindOne(context.Context, interface{}, ...*options.FindOneOptions) *mongo.SingleResult
}

type mockFinder struct {
	result []byte
}

func (mf *mockFinder) FindOne(context.Context, interface{}, ...*options.FindOneOptions) *mongo.SingleResult {
	return mongo.NewSingleResultFromBytes(mf.result)
}

func (mf *mockFinder) Find(context.Context, interface{}, ...*options.FindOptions) (*mongo.Cursor, error) {
	return &mongo.Cursor{Current: mf.result}, nil
}

func test() {
	b, err := bson.Marshal(bson.D{{"my", "value"}})
	if err != nil {
		panic(err)
	}
	mf := &mockFinder{result: b}

	doFindOne(mf)
	doFind(mf)
}
```

Running `FindOne` and decoding the `SingleResult` behaves as expected:
```go
func doFindOne(f finder) {
	var res bson.D
	err := f.FindOne(context.TODO(), bson.D{}).Decode(&res)
	if err != nil {
		panic(err)
	}
	fmt.Println(res)
}
```
Prints `[{my value}]`.

Running `Find` and iterating the cursor doesn't:
```go
func doFind(f finder) {
	cursor, err := f.Find(context.TODO(), bson.D{})
	if err != nil {
		panic(err)
	}

	for cursor.Next(context.TODO()) {
		var res bson.D
		err := cursor.Decode(&res)
		if err != nil {
			panic(err)
		}
		fmt.Println(res)
	}
}
```
Panics:
```
goroutine 1 [running]:
go.mongodb.org/mongo-driver/mongo.(*Cursor).next(0xc0000a2720, 0x163e910, 0xc0000ba008, 0x1483f00, 0xc0000c6330)
        ./mongo-go-driver/mongo/cursor.go:126 +0xef
go.mongodb.org/mongo-driver/mongo.(*Cursor).Next(...)
        ./mongo-go-driver/mongo/cursor.go:81
main.doFind(0x163a6a0, 0xc0001b0030)
        ./main.go:843 +0x17f
main.test()
        ./main.go:817 +0x145
main.main()
        ./main.go:794 +0x25
exit status 2
```

To successfully mock an API that returns a `Cursor`, we still need a way to create a `Cursor` that behaves like one returned by `Find` or `Aggregate`.

Comment (matthewdale) [inline mongo/single_result.go:33]: We also need a way to allow the `SingleResult` to return errors. Add a 2nd `error` argument or consider accepting an `io.Reader` instead of a `[]byte, error` (note that would probably require refactoring how the input is used to create a `Cursor`, but it's a more standardized way to provide bytes+errors to a function, so worth consideration).

File: mongo/single_result_test.go [hunk 4]
```diff
@@ -49,3 +50,36 @@ func TestSingleResult(t *testing.T) {
...
+	assert.Equal(t, expectedDecoded, decoded,
+		"expected decoded SingleResult to be %v, got %v", expectedDecoded, decoded)
+
+	// Assert that RDR contents are set correctly after Decode.
+	assert.NotNil(t, res.rdr, "expected non-nil rdr contents")
+	assert.Equal(t, expectedDecoded, res.rdr,
+		"expected RDR contents to be %v, got %v", expectedDecoded, res.rdr)
+
+	// Assert that a call to cur.Next will return false, as there is only one batch in
+	// SingleResult Cursors created by NewSingleResultFromBytes.
```
Comment (kevinAlbs) [inline mongo/single_result_test.go:75]:
```suggestion
	// SingleResult's Cursor created by NewSingleResultFromBytes.
```

File: x/mongo/driver/batch_cursor.go [hunk 5]
```diff
@@ -184,6 +184,13 @@ func NewEmptyBatchCursor() *BatchCursor {
 	return &BatchCursor{currentBatch: new(bsoncore.DocumentSequence)}
 }
 
+// NewBatchCursorFromBytes returns a batch cursor with current batch set to the provided contents.
+func NewBatchCursorFromBytes(contents []byte) *BatchCursor {
```
Comment (kevinAlbs) [inline x/mongo/driver/batch_cursor.go:188]: `contents` is one document rather than a batch. If creating a `BatchCursor` of one document is all that is required I suggest renaming `contents` to `document.

File: mongo/single_result_test.go [hunk 6]
```diff
@@ -49,3 +50,36 @@ func TestSingleResult(t *testing.T) {
 		assert.Equal(t, ErrNoDocuments, sr.Err(), "expected error %v, got %v", ErrNoDocuments, sr.Err())
 	})
 }
+
+func TestNewSingleResultFromBytes(t *testing.T) {
+	// Mimic a FindOne response from the server.
```
Comment (kevinAlbs) [inline mongo/single_result_test.go:55]:
```suggestion
	// Mock a document in SingleResult.
```

Nit: I interpret `response` to mean the raw server response to a cursor operation like `find` or `aggregate`. For example, running `db.runCommand({find: "foo", filter: {}})` in the shell shows the raw server response:

```
{
  cursor: {
    firstBatch: [
      { _id: ObjectId("61a7cbe3f6959812549d1134"), x: 1 },
      { _id: ObjectId("61a7ce23f6959812549d1135"), x: 2 }
    ],
    id: Long("0"),
    ns: 'test.foo'
  },
  ok: 1,
  '$clusterTime': {
    clusterTime: Timestamp({ t: 1638386659, i: 2 }),
    signature: {
      hash: Binary(Buffer.from("0000000000000000000000000000000000000000", "hex"), 0),
      keyId: Long("0")
    }
  },
  operationTime: Timestamp({ t: 1638386659, i: 2 })
}
```

That response is the expected input to `NewCursorResponse`. I suggest renaming `findOneResponse` to `findOneResult` and `findOneResponseBytes` to `findOneResultBytes`.

**Comment by benjirewis** [comment]
I’ll address your main concern first, @SVilgem:

> The main concern I have about your position is that your response for such request is always different, probably depends on who's is asking.

We regularly reconsider requests from the community and try to accommodate users when there is a common problem that we can address in an API-compatible way. We’re quite firm on the interface issue, no matter the source of the request. We had hoped this PR would be a helpful workaround for any user trying to mock the result of a read operation. The `SingleResult` type contains a number of unexported fields, so mocking a response to something like `FindOne` is not currently possible, as you cannot set or modify the fields of a `SingleResult`. This was just a suggested improvement to make mocking the result of read operations feasible without converting all our structs to interfaces.

> that’s true, but the main reason of having the interfaces is creating mocks and using them for testing.

That also sounds correct to me. However, our current adherence to semantic versioning for the Go driver does not distinguish between users’ production code and users’ testing code. So, breaking testing/mocking code is equally problematic as breaking production code. We’d really like our exposed API to be as stable as possible. Furthermore, and perhaps more importantly, it’s generally bad practice to define interfaces on the implementer side of an API for mocking. See official Golang advice [here](https://github.com/golang/go/wiki/CodeReviewComments#interfaces).

> In this case you must to bump new major release anyway…

Correct; that’s why we try to avoid removing existing functions and changing function signatures.

> Actually not only those 3, but all other as well…

If we were readily able to change structs to interfaces, I would agree that we should include all the ones you’ve included in your library (which is a very cool API, by the way).

In short, the limitations of SemVer and Golang best practices do restrict us here. Your `mongoifc` library is inspiring, though, and I’ve opened a ticket to investigate alternative solutions for mocking Go driver types. Please leave further feedback in GODRIVER-2249.

**Comment by SVilgelm** [comment]
> In short, the limitations of SemVer and Golang best practices do restrict us here.

thank you very much for the explanation and I'm sorry, I was a bit aggressive, had to solve the testing issue so many times, that it is really hard to be patient in this topic :)

> If we were readily able to change structs to interfaces, I would agree that we should include all the ones you’ve included in your library (which is a very cool API, by the way).

I tried to keep it as much as possible close to the mongo interfaces. The internal implementation has some custom hooks for logging and measuring a time, but I deleted everything from public version.

File: mongo/single_result.go (hunk 1 above)
Comment (benjirewis) [reply mongo/single_result.go:33]: Yep, we definitely need a way of mocking `Cursor` behavior. I've added a `NewMockCursor` constructor.

Comment (benjirewis) [reply mongo/single_result.go:33]: Changed `NewMockSingleResult` and `NewMockCursor` to take in an `error` and a `bsoncodec.Registry`.

File: mongo/single_result.go (hunk 2 above)
Comment (benjirewis) [reply mongo/single_result.go:36]: Changed `NewMockSingleResult` and `NewMockCursor` to take in a `bsoncodec.Registry`.

File: mongo/single_result_test.go [hunk 7]
```diff
@@ -49,3 +50,36 @@ func TestSingleResult(t *testing.T) {
 		assert.Equal(t, ErrNoDocuments, sr.Err(), "expected error %v, got %v", ErrNoDocuments, sr.Err())
 	})
 }
+
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply mongo/single_result_test.go:53]: Great idea! I believe that qualifies as an integration test, as we're technically operating against the server in your first `TODO`. I've added a new `mock_find_test.go` under `mongo/integration`. It implements something similar to the `finder` example and your code here.

File: mongo/single_result_test.go (hunk 6 above)
Comment (benjirewis) [reply mongo/single_result_test.go:55]: Great point. I've tried to use the word `result` instead of `response`, as I agree with your distinction.

File: mongo/single_result_test.go (hunk 4 above)
Comment (benjirewis) [reply mongo/single_result_test.go:75]: Done.

File: x/mongo/driver/batch_cursor.go (hunk 5 above)
Comment (benjirewis) [reply x/mongo/driver/batch_cursor.go:188]: This argument for `NewMockBatchCursor`, `NewMockCursor` and `NewMockSingleResult` is now called `documents`. For `NewMockBatchCursor`, it's a byte slice that represents the underlying data of the document sequence, so I believe it can represent a sequence-style of _multiple_ documents.

**Review by benjirewis — COMMENTED** [review]
I've done some refactoring. I'm now exposing two functions: `NewMockSingleResult` and `NewMockCursor` (note the new names) that take in a `documents` parameter of type `[]interface{}`. I think `[]interface{}` will be more convenient than `[]byte` for users; it's the type we accept for functions like [`InsertMany`](https://github.com/mongodb/mongo-go-driver/blob/master/mongo/collection.go#L235), and users will not have to use `MarshalAppend` to create a sequence-style document sequence byte slice themselves. Note that both new functions now return `error`s due to internal `Marshal`ing.

I've added unit tests in `single_result_test.go` and `cursor_test.go` for `NewMockSingleResult` and `NewMockCursor` respectively. And, I've added a new `mock_find_test.go` under `mongo/integration` to verify that a `mongo.Collection`'s behavior can be emulated using the new functions. `mock_find_test.go` can also serve as an example of how to mock read operation results.

File: mongo/integration/mock_find_test.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	docs     []interface{}
+	err      error
+	registry *bsoncodec.Registry
+}
+
+// FindOne mocks a findOne operation using NewMockSingleResult.
+func (mf *mockFinder) FindOne(_ context.Context, _ interface{}, _ ...*options.FindOneOptions) *mongo.SingleResult {
+	res, err := mongo.NewMockSingleResult(mf.docs, mf.err, mf.registry)
+
+	// If there was an error in SingleResult creation, override the error of SingleResult.
```
Comment (benjirewis) [inline mongo/integration/mock_find_test.go:38]: Mocking `FindOne` is slightly difficult here, as `NewMockSingleResult` _can_ return an error. I've chosen to override the `SingleResult.err` with the creation error if it exists, but ideally users would have access to both the original `mf.err` and the creation error.

File: mongo/cursor.go [hunk 9]
```diff
@@ -67,6 +68,45 @@ func newEmptyCursor() *Cursor {
 	return &Cursor{bc: driver.NewEmptyBatchCursor()}
 }
 
+// NewMockCursor creates a new Cursor pre-loaded with the provided documents, error and registry. If no registry is provided,
```
Comment (kevinAlbs) [inline mongo/cursor.go:71]: Nit: Though the only known use case is mocking. I suggest avoiding `Mock` in the API. Perhaps this could be named `NewCursorFromDocuments` or similar?

My concern is that we may later have use case of creating a cursor from a slice of documents, and using `NewMockCursor` seems odd.

File: mongo/cursor.go [hunk 10]
```diff
@@ -67,6 +68,45 @@ func newEmptyCursor() *Cursor {
...
+// bson.DefaultRegistry will be used.
+//
+// The documents parameter must be a slice of documents.
```
Comment (kevinAlbs) [inline mongo/cursor.go:74]: I suggest noting requirements of `documents`, similar to the documentation for `collection#InsertMany`.
```suggestion
// The documents parameter must be a slice of documents. The slice may be nil or empty.
// The elements must all be non-nil.
```

File: mongo/cursor.go [hunk 11]
```diff
@@ -67,6 +68,45 @@ func newEmptyCursor() *Cursor {
...
+func NewMockCursor(documents []interface{}, err error, registry *bsoncodec.Registry) (*Cursor, error) {
```
Comment (kevinAlbs) [inline mongo/cursor.go:75]: Is there a risk that we will need to add options in the future?

If so, we could consider adding a new `CursorOptions` struct to contain the registry:

```
func NewCursor(documents []interface{}, err error, ...options.CursorOptions)
```

That may save another function in case we need to add options in the future.

File: mongo/single_result.go [hunk 12]
```diff
@@ -29,20 +28,25 @@ type SingleResult struct {
...
-// NewSingleResultFromBytes creates a SingleResult with an underlying Cursor pre-loaded with the provided contents.
-func NewSingleResultFromBytes(contents []byte) *SingleResult {
-	c := &Cursor{
-		bc:       driver.NewBatchCursorFromBytes(contents),
-		registry: bson.DefaultRegistry,
-		Current:  bson.Raw(contents),
+// NewMockSingleResult creates a SingleResult with the provided error, registry, and an underlying Cursor pre-loaded with
+// the provided documents, error and registry. If no registry is provided, bson.DefaultRegistry will be used.
+//
+// The documents parameter must be a slice of documents.
```
Comment (kevinAlbs) [inline mongo/single_result.go:34]: I suggest making `documents` one `interface{}` instead of a slice since `SingleResult represents one document or zero on error.

File: mongo/single_result.go [hunk 13]
```diff
@@ -29,20 +28,25 @@ type SingleResult struct {
 	reg *bsoncodec.Registry
 }
 
-// NewSingleResultFromBytes creates a SingleResult with an underlying Cursor pre-loaded with the provided contents.
-func NewSingleResultFromBytes(contents []byte) *SingleResult {
-	c := &Cursor{
-		bc:       driver.NewBatchCursorFromBytes(contents),
-		registry: bson.DefaultRegistry,
-		Current:  bson.Raw(contents),
+// NewMockSingleResult creates a SingleResult with the provided error, registry, and an underlying Cursor pre-loaded with
```
Comment (kevinAlbs) [inline mongo/single_result.go:31]: I suggest removing `Mock` from the name in case there are future use-cases. Perhaps this could be `NewSingleResultFromDocument`?

File: mongo/single_result.go [hunk 14]
```diff
@@ -89,11 +93,6 @@ func (sr *SingleResult) setRdrContents() error {
 	case sr.cur != nil:
 		defer sr.cur.Close(context.TODO())
 
-		// Set contents of rdr to cur.Current if non-nil. Otherwise, iterate underlying
-		// cursor.
-		if sr.rdr = sr.cur.Current; sr.rdr != nil {
-			return nil
-		}
```
Comment (kevinAlbs) [inline mongo/single_result.go]: To check, was this condition redundant with L103-104, or was this a necessary behavior change?

File: mongo/cursor.go (hunk 9 above)
Comment (benjirewis) [reply mongo/cursor.go:71]: Sounds good. Changed to `NewCursorFromDocuments`.

File: mongo/cursor.go (hunk 10 above)
Comment (benjirewis) [reply mongo/cursor.go:74]: Good point; documented.

File: mongo/cursor.go (hunk 11 above)
Comment (benjirewis) [reply mongo/cursor.go:75]: Interesting question. At the moment, there's not much else you can specify on a `Cursor` besides a custom `*session.Client`. I suppose it's possible we'll add fields to `Cursor`, so a new `options.CursorOptions` struct _could_ be helpful in terms of backwards compatibility in the future. I'm not sure it's strictly necessary, though. Any thoughts, @matthewdale ?

[`driver.CursorOptions`](https://github.com/mongodb/mongo-go-driver/blob/master/x/mongo/driver/batch_cursor.go#L127-L135) already exists and specifies options for a batch cursor. We _could_ allow users to specify custom `driver.CursorOptions` for the underlying batch cursor created with `NewBatchCursorFromDocuments`. I'm not convinced we want to give users the ability to specify things like custom `Crypt` or `ServerAPIOptions` on their `Cursor`s created through `NewCursorFromDocuments`, though.

File: mongo/single_result.go (hunk 13 above)
Comment (benjirewis) [reply mongo/single_result.go:31]: Changed to `NewSingleResultFromDocument`.

File: mongo/single_result.go (hunk 12 above)
Comment (benjirewis) [reply mongo/single_result.go:34]: Sounds good. I do now require that the document is non-nil.

File: mongo/single_result.go (hunk 14 above)
Comment (benjirewis) [reply mongo/single_result.go]: Good catch. I added this snippet because I thought the RDR contents were not being updated correctly with `Cursor`s made with `NewCursorFromDocuments`. The code actually works as-is, and I just needed to correctly set the underlying batch of the `Cursor` [here](https://github.com/mongodb/mongo-go-driver/pull/806/files#diff-55ebb68f028f8f8bde882a1bf09524690d2b5fd42b8a04ccaa8969d39f5483b8R103-R106), so this change is no longer needed.

File: mongo/integration/mock_find_test.go (hunk 8 above)
Comment (kevinAlbs) [reply mongo/integration/mock_find_test.go:38]: Should L40 be `res.err = err`?

File: mongo/cursor.go (hunk 11 above)
Comment (kevinAlbs) [reply mongo/cursor.go]: I do not think `driver.CursorOptions` can be exposed directly.

My vote goes for adding a `CursorOptions` struct. I think it is likely cursors will change behavior in the future. Comments in DRIVERS-1857 note the possibility of implementing exhaust cursors in all drivers.

**Review by kevinAlbs — COMMENTED** [review]
Changes LGTM pending the open thread about a `CursorOptions` struct.

File: mongo/cursor.go [hunk 15]
```diff
@@ -67,6 +68,47 @@ func newEmptyCursor() *Cursor {
 	return &Cursor{bc: driver.NewEmptyBatchCursor()}
 }
 
+// NewCursorFromDocuments creates a new Cursor pre-loaded with the provided documents, error and registry. If no registry is provided,
+// bson.DefaultRegistry will be used.
+//
+// The documents parameter must be a slice of documents. The slice may be nil or empty, but all elements must be non-nil.
+func NewCursorFromDocuments(documents []interface{}, err error, registry *bsoncodec.Registry) (*Cursor, error) {
```
Comment (matthewdale) [inline mongo/cursor.go:75]: Using a `[]interface{}` to pass many documents is the same as the `InsertMany` API, but it can be difficult to use and frequently requires copying collections of documents into an `[]interface{}`. Would it be better to make the `documents` param an `interface{}` and use reflection to determine if the type is a slice or keep the API that's similar to `InsertMany`?

File: mongo/single_result.go [hunk 16]
```diff
@@ -71,6 +95,7 @@ func (sr *SingleResult) setRdrContents() error {
 		return nil
 	case sr.cur != nil:
 		defer sr.cur.Close(context.TODO())
+
 		if !sr.cur.Next(context.TODO()) {
```
Comment (matthewdale) [inline mongo/single_result.go:99]: Not part of your PR, but it's weird that we use `context.TODO()` here. We should replace those with either `context.Background()` to maintain the behavior of `context.TODO()` or use the context provided to `FindOne` in a subsequent PR.

File: x/mongo/driver/batch_cursor.go [hunk 17]
```diff
@@ -184,6 +184,15 @@ func NewEmptyBatchCursor() *BatchCursor {
 	return &BatchCursor{currentBatch: new(bsoncore.DocumentSequence)}
 }
 
+// NewBatchCursorFromDocuments returns a batch cursor with current batch set to a sequence-style
+// DocumentSequence containing the provided documents.
+func NewBatchCursorFromDocuments(documents []byte) *BatchCursor {
+	return &BatchCursor{currentBatch: &bsoncore.DocumentSequence{
+		Data:  documents,
+		Style: bsoncore.SequenceStyle,
```
Comment (matthewdale) [inline x/mongo/driver/batch_cursor.go:192]: Consider adding `id: 0` and/or `server: nil` with a comment that indicates the zero values are important because they indicate the cursor is closed and prevent the Cursor from making any actual `getMore` calls.

**Review by matthewdale — APPROVED** [review]
A few suggestions, but looks great! Thanks for leading this complex PR 👍

File: mongo/cursor.go (hunk 11 above)
Comment (matthewdale) [reply mongo/cursor.go]: I'm hesitant to add an options struct for options that we don't know exist yet. Additionally, making the mock cursor behave exactly the same as a real cursor in all circumstances is explicitly a non-goal because it's impractical (i.e. use a real MongoDB server if you need that).

Additionally, the API change you suggested can be backward compatible if added like:
```
 func NewCursorFromDocuments([]interface{}, error, *bsoncodec.Registry, ...options.CursorOptions) (*Cursor, error)
```
That is a moderately wonky API, but would facilitate the change if necessary.

File: mongo/integration/mock_find_test.go (hunk 8 above)
Comment (benjirewis) [reply mongo/integration/mock_find_test.go:38]: Ah great catch. You can't actually access `res.err` since it's an unexported field. I had to add some logic to `NewSingleResultFromDocument` to store potential creation errors on the returned `SingleResult` (noted this behavior in the documentation, too). Let me know what you think.

File: mongo/single_result.go (hunk 16 above)
Comment (benjirewis) [reply mongo/single_result.go:99]: Great point. This also happens in `topology.go`. Filed GODRIVER-2256.

File: x/mongo/driver/batch_cursor.go (hunk 17 above)
Comment (benjirewis) [reply x/mongo/driver/batch_cursor.go:192]: Good point. Added both and a relevant comment.

File: mongo/cursor.go (hunk 11 above)
Comment (kevinAlbs) [reply mongo/cursor.go]: Good points. I relent. No options seems OK given that (1) even if cursors are extended, we do not necessarily need to extend the mock API and (2) we could add a method like ApplyOptions to Cursor if we really needed.

File: mongo/cursor.go (hunk 15 above)
Comment (benjirewis) [reply mongo/cursor.go:75]: Ah interesting point. Discussed offline: it does seem that accepting an `interface{}` would be an API improvement, but the reflection required to turn an `interface{}` into its underlying `[]interface{}` would be a performance hit within the driver here. Given the precedence of the `InsertMany` API, I think it's ok to keep this as-is.

File: mongo/cursor.go (hunk 11 above)
Comment (benjirewis) [reply mongo/cursor.go]: Sounds good to me.

**Comment by cipriantarta** [comment]
Any ETA for when this gets released?

**Comment by benjirewis** [comment]
Should be very soon (next couple days) in the upcoming [1.9.0](https://jira.mongodb.org/projects/GODRIVER/versions/32094) version of the driver, @cipriantarta .
